File: java/demo/DemoServer.java

import java.io.File;
import java.util.*;

import com.google.protobuf.*;

import com.deltavsoft.rcfproto.*;

class SearchServiceImpl extends Proto.SearchService
{
    private void printRequest(
        com.google.protobuf.RpcController controller, 
        com.google.protobuf.Message request)
    {
        System.out.println("************************");
        System.out.println("*** Received request ***");

        RcfProtoController rcfController = (RcfProtoController)controller;
        RcfProtoSession session = rcfController.getSession();

        System.out.println("Client username: "        + session.getClientUsername());
        System.out.println("Compression enabled: "    + session.getEnableCompression());
        System.out.println("Transport type: "         + session.getTransportType());
        System.out.println("Transport protocol: "     + session.getTransportProtocol());

        String strRequest = request.toString();
        System.out.println("Request:");
        System.out.println(strRequest);

    }
    
    private void printResponse(com.google.protobuf.Message response)
    {
        System.out.println("*** Sending response ***");
        String strResponse = response.toString();
        System.out.println(strResponse);
        System.out.println("************************");     
    }
    
    public void search(
        com.google.protobuf.RpcController controller,
        Proto.SearchRequest request,
        com.google.protobuf.RpcCallback<Proto.SearchResponse> done)
    {       
        printRequest(controller, request);      

        // Build the response
        Proto.SearchResponse.Result result = Proto.SearchResponse.Result.newBuilder()
            .setUrl("http://someurl.com")
            .build();

        Proto.SearchResponse response = Proto.SearchResponse.newBuilder()
            .addResult(result)
            .build();
        
        printResponse(response);
        
        // Send the response.
        done.run(response);
    }


    public static boolean ShouldShutdown = false;

    public void shutdownServer(
        com.google.protobuf.RpcController controller,
        Proto.ShutdownServerRequest request,
        com.google.protobuf.RpcCallback<Proto.ShutdownServerResponse> done)
    {
        printRequest(controller, request);

        ShouldShutdown = true;
        Proto.ShutdownServerResponse response = Proto.ShutdownServerResponse.newBuilder().build();
        
        printResponse(response);
        
        done.run(response);
    }
}

public class DemoServer {

    public static void main(String[] args) {

        try
        {
            // Which OS are we on? 
            String runningOnOs = System.getProperty("os.name");
            boolean onWindows = false;
            if (runningOnOs.startsWith("Windows"))
            {
                onWindows = true;
            }

            // Configure DLL loading paths for RCFProto_Java_impl.dll / libRCFProto_Java_impl.so.
            DllPathResolver.resolve(args);
            
            // Locate path to certificates folder in distribution.
            String rootDir = DllPathResolver.findDistributionRootDir();
            File certsDir = new File(rootDir, "certs");

            // Initialize RCFProto.
            RCFProto.init();
            
            // Create server.
            RcfProtoServer server = new RcfProtoServer();

            // Bind Protobuf service.
            SearchServiceImpl myService = new SearchServiceImpl();
            server.bindService(myService);

            // Add the endpoints that this server will support.
            List<Endpoint> endpoints = new ArrayList<Endpoint>();
            endpoints.add(new TcpEndpoint(50001));
            endpoints.add(new HttpEndpoint(50002));
            
            if (onWindows)
            {
                endpoints.add(new Win32NamedPipeEndpoint("DemoServerPipe"));
            }
            else
            {
                // Delete any previous instance of the local socket.
                File file = new File("DemoServerPipe");
                if (file.exists())
                {
                    file.delete();
                }
                endpoints.add(new UnixLocalEndpoint("DemoServerPipe"));
            }           

            for (Endpoint endpoint : endpoints)
            {
                server.addEndpoint(endpoint);               
            }

            // By default, the server serves clients on a single thread. Here, we are 
            // configuring the server to use a thread pool with up to 10 threads.
            ThreadPool threadPool = new ThreadPool(1, 10);
            threadPool.setThreadName("RCFProto Server");
            server.setThreadPool(threadPool);

            // Configure SSL certificate.

            if (onWindows)
            {
                // Load certificate from PFX format.
                File certPath = new File(certsDir, "certA.p12");
                Certificate cert = new PfxCertificate(certPath.getAbsolutePath(), "", "localhost");
    
                server.setCertificate(cert);
                server.setSslImplementation(SslImplementation.Schannel);
            }
            else
            {
                // Load certificate from OpenSSL PEM format.
                File certPath = new File(certsDir, "certA.pem");
                Certificate cert = new PemCertificate(certPath.getAbsolutePath(), "");

                server.setCertificate(cert);
                server.setSslImplementation(SslImplementation.OpenSsl);             
            }

            // Start the server.
            server.start();

            // Wait for shutdown.
            while (!SearchServiceImpl.ShouldShutdown)
            {
                Thread.sleep(1000);
            }
            System.out.println("Shutting down server.");

            // Stop server.
            server.stop();

            // Deinitialize RCF.
            RCFProto.deinit();
        }
        catch(Exception e)
        {
            System.out.println("DemoServer exception:");
            System.out.println(e.getMessage());
            System.exit(1);
        }

        System.exit(0);

    }

}

File: csharp/demo/DemoServer.cs

using System;
using System.IO;
using System.Collections.Generic;
using System.Threading;

using Google.ProtocolBuffers;
using DeltaVSoft.RCFProto;
using ThreadPool = DeltaVSoft.RCFProto.ThreadPool;

// SearchService implementation.
public class SearchServiceImpl : SearchService
{
    private void PrintRequest(
        Google.ProtocolBuffers.IRpcController controller, 
        Google.ProtocolBuffers.IMessage request)
    {
        System.Console.WriteLine("************************");
        System.Console.WriteLine("*** Received request ***");

        RcfProtoController rcfController = (RcfProtoController)controller;
        RcfProtoSession session = rcfController.GetSession();

        System.Console.WriteLine("Client username: "        + session.GetClientUsername());
        System.Console.WriteLine("Compression enabled: "    + session.GetEnableCompression());
        System.Console.WriteLine("Transport type: "         + session.GetTransportType());
        System.Console.WriteLine("Transport protocol: "     + session.GetTransportProtocol());

        string strRequest = request.ToString();
        System.Console.WriteLine("Request:");
        System.Console.WriteLine(strRequest);
    }

    private void PrintResponse(Google.ProtocolBuffers.IMessage response)
    {
        System.Console.WriteLine("*** Sending response ***");
        string strResponse = response.ToString();
        System.Console.WriteLine(strResponse);
        System.Console.WriteLine("************************");
    }

    public override void Search(
        Google.ProtocolBuffers.IRpcController controller,
        SearchRequest request,
        System.Action<SearchResponse> done)
    {
        PrintRequest(controller, request);

        // Build the response.
        SearchResponse.Types.Result result = SearchResponse.Types.Result.CreateBuilder()
            .SetUrl("http://someurl.com")
            .Build();

        SearchResponse response = SearchResponse.CreateBuilder()
            .AddResult(result)
            .Build();

        PrintResponse(response);

        // Send the response.
        done(response);
    }

    public static AutoResetEvent ShutdownEvent = new AutoResetEvent(false);

    public override void ShutdownServer(
        Google.ProtocolBuffers.IRpcController controller,
        ShutdownServerRequest request,
        System.Action<ShutdownServerResponse> done)
    {
        PrintRequest(controller, request);

        ShutdownEvent.Set();
        ShutdownServerResponse response = ShutdownServerResponse.CreateBuilder().Build();

        PrintResponse(response);

        done(response);
    }

}

class Program
{
    static int Main(string[] args)
    {
        try
        {
            // Configure DLL loading paths for RCFProto_NET.dll and RCFProto_NET_impl.dll.
            DllPathResolver.Resolve(args);

            return Run();
        }
        catch (Exception e)
        {
            System.Console.WriteLine(e.Message);
        }

        return 1;
    }

    static int Run()
    {
        // Locate path to certificates folder in distribution.
        string rootDir = DllPathResolver.FindDistributionRootDir();
        string certsDir = Path.Combine(rootDir, "certs");

        // Initialize RCF.
        RCFProto.Init();

        // Create server.
        RcfProtoServer server = new RcfProtoServer();

        // Bind Protobuf service.
        SearchServiceImpl myService = new SearchServiceImpl();
        server.BindService(myService);

        // Add the endpoints that this server will support.
        List<Endpoint> endpoints = new List<Endpoint>();
        endpoints.Add(new TcpEndpoint(50001));
        endpoints.Add(new HttpEndpoint(50002));
        endpoints.Add(new Win32NamedPipeEndpoint("DemoServerPipe"));

        foreach(Endpoint endpoint in endpoints)
        {
            server.AddEndpoint(endpoint);
        }

        // By default, the server serves clients on a single thread. Here, we are 
        // configuring the server to use a thread pool with up to 10 threads.
        ThreadPool threadPool = new ThreadPool(1, 10);
        threadPool.SetThreadName("RCFProto Server");
        server.SetThreadPool(threadPool);

        // Configure SSL certificate.

        // Load certificate from PFX format.
        string certPath = Path.Combine(certsDir, "certA.p12");
        Certificate cert = new PfxCertificate(certPath, "", "localhost");

        server.SetCertificate(cert);
        server.SetSslImplementation(SslImplementation.Schannel);

        // Start the server.
        server.Start();

        // Wait for shutdown.
        SearchServiceImpl.ShutdownEvent.WaitOne();
        System.Console.WriteLine("Shutting down server.");

        // Stop server.
        server.Stop();

        return 0;
    }
}
